p2p/ipld/read.go

```diff
@@ -0,0 +1,342 @@
+package ipld
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
+)
+
+// NOTE: I'm making the assumption that bandwidth is more expensive than
+// compute, this means that when retrieving data I'm stopping as soon as enough
+// data is retrieved to recompute
+
+// ValidateAvailability implements the protocol described in https://fc21.ifca.ai/papers/83.pdf.
+// Specifically all steps of the protocol described in section
+// _5.2 Random Sampling and Network Block Recovery_ are carried out.
+//
+// In more detail it will first create numSamples random unique coordinates.
+// Then, it will ask the network for the leaf data corresponding to these coordinates.
+// Additionally to the number of requests, the caller can pass in a callback,
+// which will be called on for each retrieved leaf with a verified Merkle proof.
+//
+// Among other use-cases, the callback can be useful to monitoring (progress), or,
+// to process the leaf data the moment it was validated.
+// The context can be used to provide a timeout.
+// TODO: Should there be a constant = lower bound for #samples
+func ValidateAvailability(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	numSamples int,
+	leafSucessCb func(namespacedleaf []byte),
+) error {
+	return nil
+}
+
+/////////////////////////////////////////
+//	Retrieve Block Data
+///////////////////////////////////////
+
+type EDSParser interface {
+	Parse(eds *rsmt2d.ExtendedDataSquare) (types.Txs, types.IntermediateStateRoots, types.Messages, error)
```

Comment:
I wonder if this shouldn't simply return `Data` (and an error) as that struct wraps all the returned elements.

Reply:
Is the interface purely introduced because we weren't sure yet how the implementation will look like? If so, we can probably remove it and make it a concrete thing only?

Reply:
It should definitely be removed. I only intended to replace it once I figured out a better way.

Reply:
Can be removed once #246 lands.